src/Storages/MergeTree/MergeTreePartsMover.cpp

```diff
@@ -240,14 +243,22 @@ MergeTreeMutableDataPartPtr MergeTreePartsMover::clonePart(const MergeTreeMoveEn
     {
         cloned_part_storage = part->makeCloneOnDisk(disk, MergeTreeData::MOVING_DIR_NAME);
     }
+    String data_part_directory = cloned_part_storage->getFullPath();
+
+    TemporaryClonedPart cloned_part;
+    cloned_part.temporary_directory_lock = data->getTemporaryPartDirectoryHolder(data_part_directory);
```

Comment:
We should hold the temp dir lock until the part is renamed to persistent name (`renameTo` in `swapClonedPart`)